kafkabp/group_consumer.go

```diff
@@ -0,0 +1,140 @@
+//
+// Consumers created by NewGroupConsumer are for exactly-once message delivery
+// (e.g. stream processing a message queue). For other uses, see NewConsumer.
+//
+// All instances from the same group must share the same ClientID in the config.
+func NewGroupConsumer(cfg ConsumerConfig) (Consumer, error) {
+	sc, err := cfg.NewSaramaConfig()
+	if err != nil {
+		return nil, err
+	}
+	// Unset offset as we don't use it in group consumers.
+	sc.Consumer.Offsets.Initial = 0
+
+	consumer, err := sarama.NewConsumerGroup(cfg.Brokers, cfg.ClientID, sc)
+	if err != nil {
+		return nil, err
+	}
+	return &groupConsumer{
+		consumer: consumer,
+		cfg:      cfg,
+	}, nil
+}
+
+func (gc *groupConsumer) Consume(
+	messagesFunc ConsumeMessageFunc,
+	errorsFunc ConsumeErrorFunc,
+) error {
+	defer atomic.StoreInt64(&gc.consumeReturned, 1)
+	gc.wg.Add(1)
+	defer gc.wg.Done()
+
+	gc.wg.Add(1)
+	go func() {
+		defer gc.wg.Done()
+		for err := range gc.consumer.Errors() {
+			errorsFunc(err)
+		}
+	}()
+
+	handler := groupConsumerHandler{
+		callback: messagesFunc,
+		topic:    gc.cfg.Topic,
+	}
+
+	// gc.consumer.Consume returns when either:
+	// - rebalance happens
+	// - Close was called
+	for atomic.LoadInt64(&gc.closed) == 0 {
+		if err := gc.consumer.Consume(
+			context.Background(),
+			[]string{gc.cfg.Topic},
+			handler,
+		); err != nil {
+			errorsFunc(fmt.Errorf("sarama.ConsumerGroup.Consume returned error: %w", err))
+		}
+	}
+
+	return nil
+}
+
+// Close closes the consumer.
+func (gc *groupConsumer) Close() error {
+	atomic.StoreInt64(&gc.closed, 1)
+
+	// wait for the Consume function to return
+	defer gc.wg.Wait()
+
+	return gc.consumer.Close()
+}
+
+func (gc *groupConsumer) IsHealthy() bool {
+	return atomic.LoadInt64(&gc.consumeReturned) == 0
+}
+
+// groupConsumerHandler is the handler for sarama group consumer.
+type groupConsumerHandler struct {
+	callback ConsumeMessageFunc
+	topic    string
+}
+
+// Setup is run at the beginning of a new session, before ConsumeClaim
+func (h groupConsumerHandler) Setup(sarama.ConsumerGroupSession) error {
+	// Do nothing.
+	return nil
+}
+
+// Cleanup is run at the end of a session, once all ConsumeClaim goroutines have exited
+func (h groupConsumerHandler) Cleanup(sarama.ConsumerGroupSession) error {
+	// Do nothing.
+	return nil
+}
+
+// ConsumeClaim must start a consumer loop of ConsumerGroupClaim's Messages().
+func (h groupConsumerHandler) ConsumeClaim(session sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
+	for m := range claim.Messages() {
+		// Wrap in anonymouse function for easier defer.
+		func() {
+			ctx := context.Background()
+			var span *tracing.Span
+			spanName := "group-consumer." + h.topic
+			ctx, span = tracing.StartTopLevelServerSpan(ctx, spanName)
+			defer func() {
+				span.FinishWithOptions(tracing.FinishOptions{
+					Ctx: ctx,
+				}.Convert())
+			}()
+
+			h.callback(ctx, m)
+			session.MarkMessage(
+				m,
+				"", // metadata
```

Comment:
I wonder if this is ever useful. I did not find anything about this being a general concept in Kafka but it's documented in sarama as:

```
	// MarkOffset marks the provided offset, alongside a metadata string
	// that represents the state of the partition consumer at that point in time. The
	// metadata string can be used by another consumer to restore that state, so it
	// can resume consumption.
```
I don't think we have any real world use for this at the moment but when I looked at the implementation I tried to figure out if it makes sense to integrate it into the API :man_shrugging: